lib/class-wp-sourced-attributes-block-parser.php

```diff
@@ -0,0 +1,200 @@
+
+	/**
+	 * Parses a document and returns a list of block structures
+	 *
+	 * When encountering an invalid parse will return a best-effort
+	 * parse. In contrast to the specification parser this does not
+	 * return an error on invalid inputs.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param string                 $document Input document being parsed.
+	 * @param WP_Block_Type_Registry $registry Block type registry from which
+	 *                                         block attributes schema can be
+	 *                                         retrieved.
+	 * @param int|null               $post_id  Optional post ID.
+	 * @return WP_Block_Parser_Block[]
+	 */
+	function parse( $document, $registry = null, $post_id = null ) {
+		if ( is_null( $registry ) ) {
+			$registry = WP_Block_Type_Registry::get_instance();
+		}
+
+		$blocks = parent::parse( $document );
+
+		foreach ( $blocks as $i => $block ) {
+			$block_type = $registry->get_registered( $block['blockName'] );
+			if ( is_null( $block_type ) || ! isset( $block_type->attributes ) ) {
+				continue;
+			}
+
+			$sourced_attributes = $this->get_sourced_attributes(
+				$block,
+				$block_type->attributes,
+				$post_id
+			);
+
+			$blocks[ $i ]['attrs'] = array_merge( $block['attrs'], $sourced_attributes );
+		}
+
+		return $blocks;
+	}
+
+	/**
+	 * Returns an array of sourced attribute values for a block.
+	 *
+	 * @param WP_Block_Parser_Block $block             Parsed block object.
+	 * @param array                 $attributes_schema Attributes of registered
+	 *                                                 block type for block.
+	 * @param int|null              $post_id           Optional post ID.
+	 * @return array                                   Sourced attribute values.
+	 */
+	function get_sourced_attributes( $block, $attributes_schema, $post_id ) {
+		$attributes = array();
+
+		foreach ( $attributes_schema as $key => $attribute_schema ) {
+			if ( isset( $attribute_schema['source'] ) ) {
+				$attributes[ $key ] = $this->get_sourced_attribute(
+					$block,
+					$attribute_schema,
+					$post_id
+				);
+			}
+		}
+
+		return $attributes;
+	}
+
+	/**
+	 * Returns a sourced attribute value for a block, for attribute type which
+	 * sources from HTML.
+	 *
+	 * @param WP_Block_Parser_Block $block            Parsed block object.
+	 * @param array                 $attribute_schema Attribute schema for
+	 *                                                individual attribute to
+	 *                                                be parsed.
+	 * @return mixed                                  Sourced attribute value.
+	 */
+	function get_html_sourced_attribute( $block, $attribute_schema ) {
+		$document = new DOMDocument();
+		try {
+			// loadHTML may log warnings for unexpected markup.
+			// phpcs:ignore
+			@$document->loadHTML( '<html><body>' . $block['innerHTML'] . '</body></html>' );
+		} catch ( Exception $e ) {
+			return null;
+		}
```

Comment:
when I have worked with `DOMDocument` I found it extremely helpful to extract the setup code into a separate function that abstracts away the noisy quirks.

```php
$document = parseHTML( $block['innerHTML'] );
if ( null === $document ) {
	return null;
}
```

it's a small thing to abstract but I find in my experience it worth it especially as we learn about the settings we need to activate with whitespace and with parse-handling.

Reply:
> when I have worked with `DOMDocument` I found it extremely helpful to extract the setup code into a separate function that abstracts away the noisy quirks.

That seems like a reasonable revision, for sure! I expect it would also make writing the tests a little nicer, since I'd not need to lump all the error cases for this function otherwise intended specifically at sourcing values.